neuros/board/stream_processor.py

```diff
@@ -0,0 +1,136 @@
+import threading
+import time
+from brainflow.board_shim import BoardShim, BrainFlowInputParams, BoardIds
+
+# Global lock for thread-safe board operations
+_board_lock = threading.Lock()
+
+
+@contextmanager
+def board_stream(board_id: int = BoardIds.SYNTHETIC_BOARD) -> Iterator[BoardShim]:
+    """
+    Context manager for safe board streaming operations.
+
+    Usage:
+        with board_stream() as board:
+            data = board.get_current_board_data(100)
+    """
+    board = BoardShim(board_id, BrainFlowInputParams())
+    try:
+        board.prepare_session()
+        board.start_stream()
+        yield board
+    finally:
+        with _board_lock:
+            if board.is_prepared():
+                board.stop_stream()
+                board.release_session()
+
+
+@dataclass
+class ProcessorConfig:
+    """Configuration for the stream processor"""
+    window_size_ms: float = 50.0  # Size of processing window in milliseconds
+    overlap_ms: float = 0.0  # Overlap between windows in milliseconds
+
+
+class StreamProcessor:
+    """
+    Processes streaming data in a separate thread.
+
+    Usage:
+        def process_data(data: np.ndarray):
+            # data shape is (channels, samples)
+            print(f"Processing {data.shape[1]} samples")
+
+        with board_stream() as board:
+            processor = StreamProcessor(board)
+            processor.start(process_data)
+            time.sleep(5)  # Process for 5 seconds
+    """
+
+    def __init__(self, board: BoardShim, config: Optional[ProcessorConfig] = None):
+        self.board = board
+        self.config = config or ProcessorConfig()
+
+        # Convert ms to samples
+        sample_rate = BoardShim.get_sampling_rate(board.get_board_id())
+        self.window_samples = int(sample_rate * self.config.window_size_ms / 1000)
+        self.overlap_samples = int(sample_rate * self.config.overlap_ms / 1000)
+
+        # Processing thread management
+        self._process_thread: Optional[threading.Thread] = None
+        self._running = False
+        self._on_data: Optional[Callable[[np.ndarray], Any]] = None
+
+    def start(self, on_data: Callable[[np.ndarray], Any]) -> None:
+        """Start processing data from the stream."""
+        if self._process_thread is not None:
+            return
+
+        self._on_data = on_data
+        self._running = True
+        self._process_thread = threading.Thread(target=self._process_loop)
+        self._process_thread.daemon = True
+        self._process_thread.start()
+
+    def stop(self) -> None:
+        """Stop processing and clean up."""
+        self._running = False
+        if self._process_thread is not None:
+            self._process_thread.join()
+            self._process_thread = None
+
+    def _process_loop(self) -> None:
+        """Main processing loop."""
+        while self._running:
+            try:
+                # Get latest data window - thread-safe
+                with _board_lock:
```

Comment:
Why does reading data from the board need a lock?

Reply:
This was a confused attempt to address the potential memory leak issue, didn't document well enough and confused myself and then rushed to PR. Sorry, cleaning up as part of solution to that other problem.

Reply:
Yes, there's already a separate thread from BrainFlow that starts with the board.start_session() call, good enough for reading.